ledger/participant-integration-api/src/main/scala/platform/store/cache/StateCache.scala

```diff
@@ -73,23 +96,32 @@ private[platform] case class StateCache[K, V](cache: Cache[K, V], registerUpdate
     * (i.e. it has `validAt` highest amongst the competing updates).
     *
     * @param key the key at which to update the cache
-    * @param validAt ordering discriminator for pending updates for the same key
-    * @param eventualValue the eventual result signaling successful enqueuing of the cache async update
+    * @param fetchAsync fetches asynchronously the value for key `key` at the current cache index
     */
-  final def putAsync(key: K, validAt: Long, eventualValue: Future[V])(implicit
+  final def putAsync(key: K, fetchAsync: Long => Future[V])(implicit
       loggingContext: LoggingContext
-  ): Future[Unit] = Timed.value(
+  ): Future[V] = Timed.value(
     registerUpdateTimer,
     pendingUpdates.synchronized {
+      val validAt = _cacheIndex
+      val eventualValue = Future.delegate(fetchAsync(validAt))
```

Comment:
Just as a safeguard since this is place is performance critical. This line should return immediately, regardless of what's happening underneath.

Reply:
I have to add what we discussed offline: we do not want to be defensive in our functions returning futures. I consider this an exception since it is paramount to release the lock ASAP